tests/unit/FormCheckbox.spec.js

```diff
@@ -0,0 +1,81 @@
+import { shallowMount } from '@vue/test-utils'
+import FormCheckbox from '../../src/components/FormCheckbox.vue';
+
+describe('FormCheckbox',() => {
+    let wrapper;
+    const label = 'Accept Terms';
+    beforeEach(() => {
+        wrapper = shallowMount(FormCheckbox, {
+            propsData: {
+                label: label
+            }
+        });
+    });
+
+    it('renders the component', () => {
+        expect(wrapper.find('input').attributes().type).toBe('checkbox');
+        expect(wrapper.html()).toContain(label);
+    });
+
+    it('renders all configured properties', () => {
+        const helper = "This is some text";
+        const name = "checkboxName";
+        const error = "This field has an error";
+        const toggle = true;
+        const initiallyChecked = true;
+        const readOnly = false;
+
+        wrapper.setProps({
+            label: label,
+            name: name,
+            helper: helper,
+            toggle: toggle,
+            error: error,
+            initiallyChecked: initiallyChecked,
+            disabled: readOnly
+        });
+
+        expect(wrapper.html()).toContain(label);
+        expect(wrapper.html()).toContain(helper);
+        expect(wrapper.find('input').attributes().name).toBe(name);
+        expect(wrapper.find('.custom-switch').exists()).toBe(true);
+
+        expect(wrapper.html()).toContain(error); 
+        expect(wrapper.find('input').element.checked).toBe(true);
+        expect(wrapper.vm.checked).toBe(true);
```

Comment:
When you reach into a component's internals, (any call to `vm`), you're asserting on implementation details. Tests should facilitate refactoring, which is changing code without changing behaviour.

The way this test is set up, if we do something as basic as re-name the `checked` data variable to `isCheckd`, this test will fail. Testing internal state or implementation details leads to brittle tests that add little value.

Consider the component you are testing as a black box. You pass in some input on one end (props) and check that the correct thing has been output (the HTML). Anything in between in implementation details that shouldn't be tied to your tests.

Consumers of these components should not care what the names of things are, or how they are implemented or actually work. All they care about is that when they do _X_, they expect the correct result _Y_.

There's only a handful of things you can do when testing these components:
 * Render component with props, assert the correct HTML was rendered.
 * Trigger an action on the component (e.g. click a button), assert either:
   * The expected event was fired.
   * The expected change was made to the HTML.

Doing anything more, and you should be asking yourself questions like:
 * Am I testing internal details?
 * Will this test make refactoring difficult or impossible?
 * Does this test add any value?

Reply:
Thanks for the feedback! 

This is the first time I am writing tests, so please forgive me for any ignorance, however, I do have some questions I would like to ask. There were a few things I noticed when writing these tests that I wasn't sure how to properly handle.

Before I got deep in writing tests I wanted and nail down the tests for one element (FormInput) and base the other elements tests off of that. So I am going to be using the `FormInput` element as an example.

In testing to ensure the correct value gets emitted on input I've tried the following:
per the [vue-test-utils docs](https://vue-test-utils.vuejs.org/api/wrapper/#setprops)

```
const wrapper = shallowMount(FormInput);
wrapper.setProps({value: 'Hello World'});
expect(wrapper.emitted().input[0]).toBe('Hello World');
```
The above test fails because the input event was not emitted. 

So I tried setting the value another way:
per the [docs](https://vue-test-utils.vuejs.org/api/wrapper/#setvalue)

```
const wrapper = shallowMount(FormInput);
wrapper.find('input').setValue('Hello World');
expect(wrapper.emitted().input[0]).toBe('Hello World');
```
The above test passes, but it does not update the `value` prop on the wrapper. (At this point I wasn't sure if we even cared if the value on the wrapper was updated.)

Then using the above passing test implementation, I wrote a test to ensure the validation messages displayed when the field is invalid and was hidden when the field was valid.  
```
const wrapper = shallowMount(FormInput, {
  name: 'formInputName',
  validation: 'required'
});
expect(wrapper.find('.invalid-feedback').isVisible()).toBe(true);
expect(wrapper.html()).toContain('The formInputName field is required.');

wrapper.find('input').setValue('Hello World');
expect(wrapper.find('.invalid-feedback').isVisible()).toBe(false);
```
However, the above test fails. The error message is still visible on the front-end even though the input value is set. I noticed that if the `value` prop on the wrapper is not set, then the validation messages still appear, if the `value` is set then the messages are hidden. Realizing that we must care that the `wrapper.vm.value` is set I then based my tests to check if the `wrapper.vm.value` was updated to match the HTML element value. 

**My questions:**
1. Based off the above shouldn't the wrapper value always be set?
2. Is setting the value of the HTML element is wrong?
3. If setting the value of the HTML element is wrong, what is the best way to update the value on the wrapper? per your comment of `setProps`

> Calling setProps uses an API that will never be available to any user in the app (it changes the props in a programmatic way that I think should be reversed for special cases).
4. Am I testing this all wrong? 😆

Reply:
> ```js
> const wrapper = shallowMount(FormInput);
> wrapper.setProps({value: 'Hello World'});
> expect(wrapper.emitted().input[0]).toBe('Hello World');
> ```
> The above test fails because the input event was not emitted.

That is expected, as there's no watcher on `value` telling the component to emit when it changes. Think about how the component is used; it'll be something like this:

```html
<div class="some-container-component">
  <form-input v-model="myAwesomeValue" />
</div>
```

Above, `v-model` will bind `myAwesomeValue` to the `value` prop, and will update `myAwesomeValue` when the `input` event is fired. So when a user types into the form input, it will trigger the `input` event which will update `myAwesomeValue` in its parent, which will then update the `value` prop that is being passed down.

Without the magic of `v-model`, or setting up a wrapper with `:value` and `@input` bindings on `form-input`, there's nothing completing the loop between the `input` event and the `value` prop.

Using `setValue` seems to be the way to go.

> (At this point I wasn't sure if we even cared if the value on the wrapper was updated.)

I don't think we need to care about the `value` prop changing in this case. Whether or not the user has decided to correctly set up `v-model` or `:value`/`@change` bindings is not something that component has control over right now; all we can do is:
 * Ensure that when a user enters text, the correct event is emitted with the correct value, and
 * When the `value` prop changes (I think using `setProps` in this case is the right thing) the form input displays the value.

So a test for this might look like:
```js
it('emits value when input changes', () => {
  const wrapper = shallowMount(FormInput);
  const value = 'Some text!';

  wrapper.find('input').setValue(value);

  expect(wrapper.emitted('input')[0]).toEqual([value]);
});

it('sets value on input', () => {
  const wrapper = shallowMount(FormInput);
  const value = 'Some text!';

  wrapper.setProps({ value });

  expect(wrapper.find('input').element.value).toEqual(value);
});
```

Regarding your questions:

> 1. Based off the above shouldn't the wrapper value always be set?

I don't think so. It's possible for the component to be mounted and for any value to be set, as there's no prop validation. I do think that it might be worth testing with mounting with an initial value, then calling setProps, and also mounting with no value and asserting that the input is indeed empty.

> 2. Is setting the value of the HTML element is wrong?

Nope. Sorry if I confused you here! I was thinking in general terms and how some user might use the component. But in this case, we're unit testing a collection of library components, so there's no user; a consumer would be another component that would actually be setting props on the child.

A component has a simple public API: props down, events up, and should be tested using that API. Setting props is perfectly valid in this case. I think my advice on avoiding `setProps` makes more sense for nested components (which may come up if doing more high level integration or e2e tests).

Reply:
Regarding the validation: It seems `isVisible` checks for `display: none` or `visibility: hidden`, not specifically if the element doesn't exist at all (as is the case when using `v-if`). Maybe try testing using [`exists`](https://vue-test-utils.vuejs.org/api/wrapper/#exists)?